enterprise/suppress-user/syncer.go

```diff
@@ -0,0 +1,227 @@
+	"github.com/rudderlabs/rudder-server/enterprise/suppress-user/model"
+	"github.com/rudderlabs/rudder-server/services/controlplane/identity"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/rudderlabs/rudder-server/utils/types/deployment"
+)
+
+// SyncerOpt represents a configuration option for the syncer
+type SyncerOpt func(*Syncer)
+
+// WithHttpClient sets the http client to use
+func WithHttpClient(client *http.Client) SyncerOpt {
+	return func(c *Syncer) {
+		c.client = client
+	}
+}
+
+// WithPageSize sets the page size for each sync request
+func WithPageSize(pageSize int) SyncerOpt {
+	return func(c *Syncer) {
+		c.pageSize = pageSize
+	}
+}
+
+// WithPollIntervalFn sets the interval at which the syncer will poll the backend
+func WithPollIntervalFn(pollIntervalFn func() time.Duration) SyncerOpt {
+	return func(c *Syncer) {
+		c.pollIntervalFn = pollIntervalFn
+	}
+}
+
+// WithLogger sets the logger to use in the syncer
+func WithLogger(log logger.Logger) SyncerOpt {
+	return func(c *Syncer) {
+		c.log = log
+	}
+}
+
+// MustNewSyncer creates a new syncer, panics if an error occurs
+func MustNewSyncer(baseURL string, identifier identity.Identifier, r Repository, opts ...SyncerOpt) *Syncer {
+	s, err := NewSyncer(baseURL, identifier, r, opts...)
+	if err != nil {
+		panic(err)
+	}
+	return s
+}
+
+// NewSyncer creates a new syncer
+func NewSyncer(baseURL string, identifier identity.Identifier, r Repository, opts ...SyncerOpt) (*Syncer, error) {
+	var url string
+	switch identifier.Type() {
+	case deployment.DedicatedType:
+		url = fmt.Sprintf("%s/dataplane/workspaces/%s/regulations/suppressions", baseURL, identifier.ID())
+	case deployment.MultiTenantType:
+		url = fmt.Sprintf("%s/dataplane/namespaces/%s/regulations/suppressions", baseURL, identifier.ID())
+	default:
+		return nil, fmt.Errorf("unsupported deployment type: %s", identifier.Type())
+	}
+
+	s := &Syncer{
+		url:                url,
+		r:                  r,
+		log:                logger.NOP,
+		client:             &http.Client{},
+		pageSize:           100,
+		pollIntervalFn:     func() time.Duration { return 30 * time.Second },
+		defaultWorkspaceID: identifier.ID(),
+	}
+	for _, opt := range opts {
+		opt(s)
+	}
+	return s, nil
+}
+
+// Syncer is responsible for syncing suppressions from the backend to the repository
+type Syncer struct {
+	url string
+	r   Repository
+
+	client             *http.Client
+	log                logger.Logger
+	pageSize           int
+	pollIntervalFn     func() time.Duration
+	defaultWorkspaceID string
+}
+
+// SyncLoop runs the sync loop until the provided context is done
+func (s *Syncer) SyncLoop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(s.pollIntervalFn()):
+		}
+	again:
+		s.log.Info("Fetching Regulations")
+		token, err := s.r.GetToken()
+		if err != nil {
+			if errors.Is(err, model.ErrRestoring) {
+				time.Sleep(1 * time.Second)
```

Comment:
You should sleep by respecting the context. There is a `misc.SleepCtx` function if you need it.